File: practice/src/main/java/com/decade/practice/web/rest/UserController.java
package com.decade.practice.web.rest;

import com.decade.practice.data.repositories.UserRepository;
import com.decade.practice.media.ImageStore;
import com.decade.practice.model.domain.DefaultAvatar;
import com.decade.practice.model.domain.embeddable.ImageSpec;
import com.decade.practice.model.domain.entity.User;
import com.decade.practice.model.dto.SignUpRequest;
import com.decade.practice.security.model.DaoUser;
import com.decade.practice.usecases.UserOperations;
import com.decade.practice.utils.ImageUtils;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.context.SecurityContextRepository;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.net.URI;
import java.util.Collections;
import java.util.List;

@RestController
@RequestMapping("/users")
public class UserController {

        private final UserRepository userRepository;
        private final SecurityContextRepository contextRepo;
        private final ImageStore imageStore;
        private final UserOperations userOperations;

        public UserController(UserRepository userRepository,
                              SecurityContextRepository contextRepo,
                              ImageStore imageStore,
                              UserOperations userOperations
        ) {
                this.userOperations = userOperations;
                this.userRepository = userRepository;
                this.contextRepo = contextRepo;
                this.imageStore = imageStore;
        }

        @GetMapping
        public List<User> findConversations(
                @AuthenticationPrincipal(expression = "name") String username,
                @RequestParam(required = true) String query
        ) {
                return userRepository.findByNameContainingAndRole(query, "ROLE_USER");
        }

        @PostMapping
        @PreAuthorize("isAnonymous()")
        public ResponseEntity<String> registerUser(
                HttpServletRequest request,
                HttpServletResponse response,
                @RequestPart("information") @Valid SignUpRequest information,
                @RequestPart(value = "file", required = false) MultipartFile file
        ) throws IOException {
                ImageSpec avatar;
                if (file != null) {
                        avatar = imageStore.save(ImageUtils.crop(file.getInputStream()));
                } else {
                        avatar = DefaultAvatar.getInstance();
                }

                try {
                        User user = userOperations.create(
                                information.getUsername(),
                                information.getPassword(),
                                information.getName(),
                                information.getDob(),
                                information.getGender(),
                                avatar,
                                true
                        );

                        SecurityContext context = SecurityContextHolder.createEmptyContext();
                        context.setAuthentication(new UsernamePasswordAuthenticationToken(
                                new DaoUser(user), information.getPassword(),
                                Collections.singletonList(new SimpleGrantedAuthority("ROLE_USER"))
                        ));
                        contextRepo.saveContext(context, request, response);
                        SecurityContextHolder.setContext(context);

                        return ResponseEntity.status(HttpStatus.CREATED).build();
                } catch (Exception e) {
                        e.printStackTrace();
                        if (avatar != DefaultAvatar.getInstance()) {
                                imageStore.remove(URI.create(avatar.getUri()));
                        }

                        if (e instanceof DataIntegrityViolationException) {
                                return ResponseEntity.status(HttpStatus.CONFLICT.value()).body("Username exists");
                        } else {
                                throw e;
                        }
                }
        }
}

File: practice/src/main/java/com/decade/practice/endpoints/user/UserController.kt
package com.decade.practice.endpoints.user

import com.decade.practice.database.UserOperations
import com.decade.practice.image.ImageStore
import com.decade.practice.model.embeddable.ImageSpec
import com.decade.practice.model.entity.User
import com.decade.practice.security.model.JwtUser
import com.decade.practice.util.ImageUtils
import jakarta.persistence.OptimisticLockException
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.Past
import jakarta.validation.constraints.Size
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.stereotype.Controller
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.util.*

@Controller
@RequestMapping("/user")
@Validated
class UserController(
    private val imageStore: ImageStore,
    private val userOperations: UserOperations
) {

    @PostMapping("/information")
    @Throws(OptimisticLockException::class)
    fun updateInformation(
        @AuthenticationPrincipal(expression = "id") id: UUID,
        @Size(min = 4, max = 100, message = "Name length must be between 4-100")
        @RequestParam name: String,
        @Past(message = "Do you have time machine.")
        @RequestParam dob: Date,
        @NotEmpty
        @RequestParam gender: String
    ): ResponseEntity<User> {
        return ResponseEntity.ok(userOperations.update(id, name, dob, gender))
    }

    @PostMapping("/avatar")
    @Throws(IOException::class, OptimisticLockException::class)
    fun updateAvatar(
        httpRequest: HttpServletRequest,
        @AuthenticationPrincipal jwtUser: JwtUser,
        @RequestParam file: MultipartFile
    ): ResponseEntity<User> {
        val cropped = ImageUtils.crop(file.inputStream)
        val url = imageStore.save(cropped)
        val avatar = ImageSpec(url.toString())
        try {
            return ResponseEntity.ok(userOperations.update(jwtUser.id, avatar))
        } catch (e: Exception) {
            imageStore.remove(url)
            throw e
        }
    }
}
